liblwgeom/lwgeom_transform.c

```diff
@@ -126,49 +198,268 @@ lwgeom_transform(LWGEOM *geom, projPJ inpj, projPJ outpj)
+		return LW_SUCCESS;
+
+	switch(geom->type)
 	{
-		int pj_errno_val = *pj_get_errno_ref();
-		if (pj_errno_val == -38)
+		case POINTTYPE:
+		case LINETYPE:
+		case CIRCSTRINGTYPE:
+		case TRIANGLETYPE:
 		{
-			lwnotice("PostGIS was unable to transform the point because either no grid"
-				 " shift files were found, or the point does not lie within the "
-				 "range for which the grid shift is defined. Refer to the "
-				 "ST_Transform() section of the PostGIS manual for details on how "
-				 "to configure PostGIS to alter this behaviour.");
-			lwerror("transform: couldn't project point (%g %g %g): %s (%d)",
-				orig_pt.x, orig_pt.y, orig_pt.z,
-				pj_strerrno(pj_errno_val), pj_errno_val);
+			LWLINE *g = (LWLINE*)geom;
+			if (!ptarray_transform(g->points, pj))
+				return LW_FAILURE;
+			break;
 		}
-		else
+		case POLYGONTYPE:
 		{
-			lwerror("transform: couldn't project point (%g %g %g): %s (%d)",
-				orig_pt.x, orig_pt.y, orig_pt.z,
-				pj_strerrno(pj_errno_val), pj_errno_val);
+			LWPOLY *g = (LWPOLY*)geom;
+			for (i = 0; i < g->nrings; i++)
+			{
+				if (!ptarray_transform(g->rings[i], pj))
+					return LW_FAILURE;
+			}
+			break;
+		}
+		case MULTIPOINTTYPE:
+		case MULTILINETYPE:
+		case MULTIPOLYGONTYPE:
+		case COLLECTIONTYPE:
+		case COMPOUNDTYPE:
+		case CURVEPOLYTYPE:
+		case MULTICURVETYPE:
+		case MULTISURFACETYPE:
+		case POLYHEDRALSURFACETYPE:
+		case TINTYPE:
+		{
+			LWCOLLECTION *g = (LWCOLLECTION*)geom;
+			for (i = 0; i < g->ngeoms; i++)
+			{
+				if (!lwgeom_transform(g->geoms[i], pj))
+					return LW_FAILURE;
+			}
+			break;
 		}
+		default:
+		{
+			lwerror("lwgeom_transform: Cannot handle type '%s'",
+			          lwtype_name(geom->type));
+			return LW_FAILURE;
+		}
+	}
+	return LW_SUCCESS;
+}
+
+static int
+proj_crs_is_swapped(const PJ* pj_crs)
+{
+	PJ *pj_cs = proj_crs_get_coordinate_system(NULL, pj_crs);
+	int axis_count = proj_cs_get_axis_count(NULL, pj_cs);
+	if (axis_count > 0)
+	{
+		const char *out_name, *out_abbrev, *out_direction;
+		double out_unit_conv_factor;
+		const char *out_unit_name, *out_unit_auth_name, *out_unit_code;
+		/* Read only first axis, see if it is degrees / north */
+		proj_cs_get_axis_info(NULL, pj_cs, 0,
+			&out_name,
+			&out_abbrev,
+			&out_direction,
+			&out_unit_conv_factor,
+			&out_unit_name,
+			&out_unit_auth_name,
+			&out_unit_code
+			);
+		return (strcasecmp(out_direction, "north") == 0);
+	}
+	return LW_FALSE;
+}
+
+
+int
+ptarray_transform(POINTARRAY* pa, PJ* pj)
+{
+	uint32_t i;
+	POINT4D p;
+	size_t n_converted;
+	size_t n_points = pa->npoints;
+	size_t point_size = ptarray_point_size(pa);
+	int has_z = ptarray_has_z(pa);
+	double *pa_double = (double*)(pa->serialized_pointlist);
+	int input_swapped, output_swapped;
+
+	/* XXXX TODO check that the PJ has decimal degrees as units in input/output */
+
+	PJ* pj_source_crs = proj_get_source_crs(NULL, pj);
+	PJ* pj_target_crs = proj_get_target_crs(NULL, pj);
+
+	if (!(pj_source_crs && pj_source_crs))
```

Comment:
This is checking the same pointer twice. Should it be `pj_source_crs && pj_target_crs`?
In any case, if only one of them failed the other needs to be destroyed to avoid leaks.